### DISCUSSION TITLE
Would you accept a PR to add support for Gerrit 2.14 EOL?

### DISCUSSION THREAD

**Discussion by felipecrs** [renovatebot/renovate#27668] [Suggest an Idea]
### Tell us more.

Gerrit 2.14 is EOL by some years by now, however, we use it daily at work and upgrading is not an option.

We have patched Renovate's Gerrit support to support Gerrit 2.14, which lacks some endpoints and has a few differences in other endpoints.

Adding support back to Gerrit 2.14 will also enable support for any newer version of it, of course, like 2.15 and 2.16 (the non notedb variant).

My question is: would you welcome if I contribute this back? Reason I ask before sending a PR is because while I have code to support the old Gerrit running for months, I have not yet fixed the test cases neither kept support for the newer versions, and I don't want to spend time doing that if it's not of the project's interest.

**Comment by viceice** [comment]
as long as it doesn't make the code too complicated it's ok.

but we (renovate maintainer) will never ensure to support long eol platforms. that will always need community contributions. so it can happen to break those by newer contributions, even it's unintentional.

**Reply by felipecrs** [reply]
> as long as it doesn't make the code too complicated it's ok.

Awesome! And no, I don't think it will make the code too complicated.

> but we (renovate maintainer) will never ensure to support long eol platforms. that will always need community contributions. so it can happen to break those by newer contributions, even it's unintentional.

All good, I was not expecting anything different.

PS: no promises as to when I should submit a PR, quite busy ATM.

**Comment by viceice** [comment]
btw: what the reason to not do minor updates to a supported version?

you hopefully don't allow Internet access to that old platform for security reasons.

**Reply by felipecrs** [reply]
It's a long story... to make it short:

- It's beyond my control, this is taken care by a different team
- Upgrade to 2.16 and 3.4 were attempted multiple times already, but there's always some roadblocks like performance degradation that can't be solved during the tight upgrade window they have.

Hopefully our Gerrit will be upgraded someday, but this is something that may happen this year or in the next 5 years: we can't wait for it.

**Comment by felipecrs** [comment]
Turns out our company finally upgraded Gerrit to 3.x. This means I will not send any PRs to add support for Gerrit 2.14.

But in case it is useful for you, here are the needed patches:

```patch
diff --git a/lib/modules/platform/gerrit/client.ts b/lib/modules/platform/gerrit/client.ts
index 3182db64d..3cebbf6da 100644
--- a/lib/modules/platform/gerrit/client.ts
+++ b/lib/modules/platform/gerrit/client.ts
@@ -99,9 +99,14 @@ class GerritClient {
   }
 
   async setCommitMessage(changeNumber: number, message: string): Promise<void> {
-    await this.gerritHttp.putJson(`a/changes/${changeNumber}/message`, {
+    await this.gerritHttp.putJson(`a/changes/${changeNumber}/edit:message`, {
       body: { message },
     });
+    await this.gerritHttp.postJson(`a/changes/${changeNumber}/edit:publish`, {
+      body: {
+        notify: 'NONE',
+      },
+    });
   }
 
   async updateChangeSubject(
@@ -118,11 +123,12 @@ class GerritClient {
   }
 
   async getMessages(changeNumber: number): Promise<GerritChangeMessageInfo[]> {
-    const messages = await this.gerritHttp.getJson<GerritChangeMessageInfo[]>(
-      `a/changes/${changeNumber}/messages`,
+    const change = await this.gerritHttp.getJson<GerritChange>(
+      `a/changes/${changeNumber}/detail`,
       { memCache: false },
     );
-    return messages.body;
+
+    return change.body.messages!;
   }
 
   async addMessage(
@@ -231,8 +237,10 @@ class GerritClient {
   ): string[] {
     const filterState = mapPrStateToGerritFilter(searchConfig.state);
     const filters = ['owner:self', 'project:' + repository, filterState];
-    if (searchConfig.branchName !== '') {
-      filters.push(`hashtag:sourceBranch-${searchConfig.branchName}`);
+    if (searchConfig.branchName) {
+      filters.push(`topic:source-${searchConfig.branchName}`);
+    } else {
+      filters.push('intopic:source-');
     }
     if (searchConfig.targetBranch) {
       filters.push(`branch:${searchConfig.targetBranch}`);
diff --git a/lib/modules/platform/gerrit/scm.ts b/lib/modules/platform/gerrit/scm.ts
index f4369fa39..c775771df 100644
--- a/lib/modules/platform/gerrit/scm.ts
+++ b/lib/modules/platform/gerrit/scm.ts
@@ -123,7 +123,7 @@ export class GerritScm extends DefaultGitScm {
       if (hasChanges || commit.force) {
         const pushResult = await git.pushCommit({
           sourceRef: commit.branchName,
-          targetRef: `refs/for/${commit.baseBranch!}%t=sourceBranch-${
+          targetRef: `refs/for/${commit.baseBranch!}%topic=source-${
             commit.branchName
           }`,
           files: commit.files,
diff --git a/lib/modules/platform/gerrit/types.ts b/lib/modules/platform/gerrit/types.ts
index 3aae7d47e..e55b55b25 100644
--- a/lib/modules/platform/gerrit/types.ts
+++ b/lib/modules/platform/gerrit/types.ts
@@ -34,6 +34,7 @@ export type GerritReviewersType = 'REVIEWER' | 'CC' | 'REMOVED';
 
 export interface GerritChange {
   branch: string;
+  topic?: string;
   hashtags?: string[];
   change_id: string;
   subject: string;
diff --git a/lib/modules/platform/gerrit/utils.ts b/lib/modules/platform/gerrit/utils.ts
index 75ca900d1..f220dace0 100644
--- a/lib/modules/platform/gerrit/utils.ts
+++ b/lib/modules/platform/gerrit/utils.ts
@@ -55,7 +55,7 @@ export function mapPrStateToGerritFilter(state?: PrState): string {
       return 'status:open';
     case 'all':
     default:
-      return '-is:wip';
+      return '-is:draft';
   }
 }
 
@@ -91,9 +91,10 @@ export function mapGerritChangeStateToPrState(
 }
 
 export function extractSourceBranch(change: GerritChange): string | undefined {
-  return change.hashtags
-    ?.find((tag) => tag.startsWith('sourceBranch-'))
-    ?.replace('sourceBranch-', '');
+  if (change.topic?.startsWith('source-')) {
+    return change.topic.replace('source-', '');
+  }
+  return undefined;
 }
 
 export function findPullRequestBody(change: GerritChange): string | undefined {

```

Basically, Gerrit 2.14 without NoteDB doesn't have support for hashtags, so it's refactored to use topics instead.

Also, such version of Gerrit doesn't have the REST API to set commit message. But it's worked around using some other APIs.

Also, WIP changes are a feature of newer Gerrit versions. Older versions of Gerrit had the concept of draft changes instead.

**Reply by felipecrs** [reply]
Also, it's worth mentioning that this patch will/would be much simpler after/if https://github.com/renovatebot/renovate/pull/29802 is merged. That's because the "footer" approach proposed in that PR is compatible with both old and newer versions of Gerrit.
